### ISSUE TITLE
Questions about the Raw Data

### ISSUE THREAD

**Issue by yangky11** [Simontwice/MagnusData#1]
Hi,

Thanks for sharing this great work! I have a few questions about the raw data used by this repo for generating the final datasets.

* How to generate the raw data files? Do you plan to release the code of this part?
* Does the data contain a list of all defined premises (regardless of whether they are used by proofs)? If not, what if a testing proof step requires a premise that has never been used in training?
* Given a testing theorem, do we know the subset of premises that it can access?

**Comment by Simontwice** [comment]
Hi, thank you for reaching out! We are very excited to share our work.

> Does the data contain a list of all defined premises (regardless of whether they are used by proofs)? 

The data does not contain the list of all possible premises, as some facts (lemmas, theorems etc.) proved in the source data are never used as premises in any proof. Such facts only appear in the dataset as the theorem to be proven, provided their proof utilises at least 1 premise.

> If not, what if a testing proof step requires a premise that has never been used in training?

The idea in Magnushammer is that the model should be able to generalise to previously unseen lemmas. During test time, the model is presented a large number of premises, some of them previously unseen, and is able to effectively choose the most relevant ones, based on the textual representations of the premises and the state of the proof. This is also relevant to the following:
> Given a testing theorem, do we know the subset of premises that it can access?

Correct, during test time we see all the premises available in the context of the proof. It contains lemmas/theorems, definitions, axioms etc, both local and imported from other files. It is usually 20K-30K premises.

> How to generate the raw data files? Do you plan to release the code of this part?

While we are not able to release the code at the moment, it is possible that we do so in the future. However, we would be happy to discuss this matter further and explore our options.

**Comment by yangky11** [comment]
Hi, 

Thank you for your quick response! Given a particular proof, how do we know which premises are available? Presumably, they include lemmas defined in the same file before the current proof, as well as lemmas imported from other files, but they shouldn't include lemmas after the current proofs and lemmas that are not imported?

Just to clarify, I'm wondering how to compute the set of available premises.

**Comment by Simontwice** [comment]
> Presumably, they include lemmas defined in the same file before the current proof, as well as lemmas imported from other files, but they shouldn't include lemmas after the current proofs and lemmas that are not imported?

You are absolutely correct, this is the way it is done in Magnushammer, and is in general what happens in the Isabelle interface we used. It's called [PISA](https://github.com/albertqjiang/Portal-to-ISAbelle), developed mainly by Albert Jiang.

> Just to clarify, I'm wondering how to compute the set of available premises.

It would be helpful to know the context in which you want to use the available premises, in other words: what it is you need the premises for. Is it for negative datapoints in contrastive learning? A set to choose from during eval time? The answer will likely vary, depending on that.

Very excited to aid you in using the data!

**Comment by yangky11** [comment]
Hi,

I meant the set of premises to choose from during evaluation.

**Comment by Simontwice** [comment]
Ah, I see. In this case, I think the path of least resistance would be for our team to generate the set of available premises for every theorem in the test set and include it here + on huggingface.

The other possibility would be for you to set up PISA on your machine and retrieve the premises using the API, but most of the time it is computationally demanding and may take some know-how.

How does the option of us adding the premises sound to you?

**Comment by yangky11** [comment]
Hi,

I see. Could you please point me to the PISA APIs for generating the set of available premises? I can probably take a deeper look at PISA, but I didn't see premises mentioned in PISA's README. Thank you!

**Comment by yangky11** [comment]
Hi @Simontwice , just to follow up on this. Thanks!
